### PR TITLE
[Glimmer2] Implement glimmer local lookup for components

### PR DESCRIPTION
**Pull request by asakusuma**
This PR requires this [glimmer template meta PR](https://github.com/tildeio/glimmer/pull/141) before the tests will pass.

cc @rwjblue

### CODE REVIEW

File: packages/ember-glimmer/lib/environment.js [hunk 1]
```diff
@@ -79,14 +79,20 @@ export default class Environment extends GlimmerEnvironment {
...
+      let { component: ComponentClass, layout } = lookupComponent(this.owner, name);
 
       if (ComponentClass || layout) {
-        definition = this._components[name] = new CurlyComponentDefinition(name, ComponentClass, layout);
+        definition = this._components[name] = new CurlyComponentDefinition(nameSegments, ComponentClass, layout);
+      } else if (meta && meta.moduleName) {
+        let segments = meta.moduleName.split('/')
+        segments.shift();
+        segments.push(name);
+        return this.getComponentDefinition(segments);
```
Comment (rwjblue) [inline packages/ember-glimmer/lib/environment.js]: This is slightly confusing to me, can you explain the idea here?

File: packages/ember-glimmer/lib/environment.js [hunk 2]
```diff
@@ -79,14 +79,20 @@ export default class Environment extends GlimmerEnvironment {
...
   }
 
-  getComponentDefinition(name) {
+  getComponentDefinition(nameSegments, meta) {
+    let name = nameSegments.join('/');
     let definition = this._components[name];
 
     if (!definition) {
-      let { component: ComponentClass, layout } = lookupComponent(this.owner, name[0]);
+      let { component: ComponentClass, layout } = lookupComponent(this.owner, name);
```
Comment (rwjblue) [inline packages/ember-glimmer/lib/environment.js]: To support local lookup we need to pass a third arg to `lookupComponent` here ([see here](https://github.com/emberjs/ember.js/blob/master/packages/ember-glimmer/lib/utils/lookup-component.js#L27)). I think that arg would be `{ source: meta && meta.moduleName}`.  Then that helper function looks for a local result first, and falls back to global.  I'm not sure why we need `else if` below if we pass the third arg to `lookupComponent`.

File: packages/ember-glimmer/lib/environment.js (hunk 1 above)
Comment (asakusuma) [inline packages/ember-glimmer/lib/environment.js]: moduleName will be somethings like `compoment/parent-foo` and `nameSegments` will be `['my-foo']`. This code creates `segments`, which ends up being `['parent-foo', 'my-foo']` and becomes the `nameSegments` of the recursive call.

File: packages/ember-glimmer/lib/environment.js (hunk 2 above)
Comment (asakusuma) [inline packages/ember-glimmer/lib/environment.js]: Ahh, I think I need to get rid of the code you commented on below and instead use this third parameter.

Comment (rwjblue) [inline packages/ember-glimmer/lib/environment.js]: 👍 - The idea is that `lookupComponent` is the only thing that has to care about the fallback stuff.

**Comment by asakusuma** [comment]
Closed in favor of https://github.com/emberjs/ember.js/pull/13378

**Comment by stefanpenner** [comment]
@asakusuma such :trollface: that link links to this PR, i clicked it 3 times :P

**Comment by asakusuma** [comment]
@stefanpenner you know how I do. Closed in favor of https://github.com/emberjs/ember.js/pull/13778

**Pull request closed without merging**